crates/plugins/src/manager.rs

```diff
@@ -0,0 +1,219 @@
+        if let Ok(installed) = self.store.read_plugin_manifest(&plugin_manifest.name()) {
+            if &installed == plugin_manifest {
+                bail!(
+                    "Plugin {} is already installed with version {}.",
+                    plugin_manifest.name(),
+                    installed.version,
+                );
+            } else if installed.version > plugin_manifest.version && !allow_downgrades {
+                bail!(
+                    "Newer version {} of {} plugin is already installed. To downgrade to version {} set the `--downgrade` flag.",
+                    installed.version,
+                    plugin_manifest.name(),
+                    plugin_manifest.version,
+                );
+            }
+        }
+
+        check_supported_version(plugin_manifest, spin_version)
+    }
+
+    /// Fetches a manifest from a local, remote, or repository location and returned the parsed
+    /// PluginManifest object.
+    pub async fn get_manifest(
+        &self,
+        manifest_location: &ManifestLocation,
+    ) -> Result<PluginManifest> {
+        let plugin_manifest = match manifest_location {
+            ManifestLocation::Remote(url) => {
+                log::info!("Pulling manifest for plugin from {}", url);
+                reqwest::get(url.as_ref())
+                    .await?
+                    .error_for_status()
+                    .with_context(|| format!("Manifest endpoint {} is unhealthy", url))?
+                    .json::<PluginManifest>()
+                    .await
+                    .with_context(|| format!("Failed to deserialize remote plugin at {}", url))?
+            }
+            ManifestLocation::Local(path) => {
+                log::info!("Pulling manifest for plugin from {}", path.display());
+                let file = File::open(path).with_context(|| {
+                    format!("The local manifest {} could not be opened", path.display())
+                })?;
+                serde_json::from_reader(file).with_context(|| format!("Failed to deserialize local manifest at {}. Ensure it has the expected plugin manifest format", path.display()))?
+            }
+            ManifestLocation::PluginsRepository(lookup) => {
+                lookup
+                    .get_manifest_from_repository(self.store().get_plugins_directory())
+                    .await?
+            }
+        };
+        Ok(plugin_manifest)
+    }
+}
+
+/// Gets the appropriate package for the running OS and Arch if exists
+pub fn get_package(plugin_manifest: &PluginManifest) -> Result<&PluginPackage> {
+    let os = std::env::consts::OS;
+    let arch = std::env::consts::ARCH;
+    plugin_manifest
+        .packages
+        .iter()
+        .find(|p| p.os.to_string() == os && p.arch.to_string() == arch)
+        .ok_or_else(|| {
+            anyhow!(
+                "This plugin does not support this OS ({}) or architecture ({}).",
+                os,
+                arch
+            )
```

Comment:
You could bring symbols into function-local scope instead of assigning here:

```suggestion
    use std::env::consts::{ARCH, OS};
    plugin_manifest
        .packages
        .iter()
        .find(|p| p.os.to_string() == OS && p.arch.to_string() == ARCH)
        .ok_or_else(|| {
            anyhow!(
                "This plugin does not support this OS ({OS}) or architecture ({ARCH}).")
```